### ISSUE TITLE
Would this clash with react-hot-loader?

### ISSUE THREAD

**Issue by stereokai** [caspervonb/amok#92]
Can amok work alongside [react-hot-loader](http://gaearon.github.io/react-hot-loader/)? Or would they interrupt each other on the race to update?

**Comment by caspervonb** [comment]
They will race, one sho

**Comment by stereokai** [comment]
Thanks for a super fast reply. What do you mean replicated in client side code? Run it in the browser? It is a webpack plugin. There is also a browserify equivalent.

In spite of everything, the creator has mentioned you in his latest article:
https://medium.com/@dan_abramov/the-death-of-react-hot-loader-765fa791d7c4

**Comment by caspervonb** [comment]
> Or would they interrupt each other on the race to update?

Amok would update the functions without disturbing the bindings, e.g the function instance is the same, this is what allow event handlers, callbacks and other closures to be edited.

I'm pretty sure react-hot-loader will replace the binding with a new function assignment, breaking the binding. This is what the two do different.

> Thanks for a super fast reply. What do you mean replicated in client side code?
> Run it in the browser? It is a webpack plugin. There is also a browserify equivalent.
The actual implementation for both is all client side, to make things atlas predictable I'd try to remote their server components and use amok's patch events instead.

**Comment by stereokai** [comment]
> remove their server components and use amok's patch events instead.

This could actually revolutionize current React tooling and coding speed.

**Comment by caspervonb** [comment]
Summoning @gaearon, if you have anything to add to  the topic :smile:

**Comment by stereokai** [comment]
@gaearon Danny Boy is a busy guy :)

But hopefully he will drop by and notice us. Have you seen his latest presentation on React Europe?

[![Dan Abramov - Live React: Hot Reloading with Time Travel at react-europe 2015](http://img.youtube.com/vi/xsSnOQynTHs/0.jpg)](http://www.youtube.com/watch?v=xsSnOQynTHs)

**Comment by caspervonb** [comment]
@stereokai there is room for user experimentation and contributions here tho, Just trying both together would be a start, see what happens :smile: 

I'm realising I cannot know everything about every framework, build pipeline etc out there, It's just too much and I'm already struggling hard with time.
Going forward there's going to be far less time as funding isn't going too great and likely will be halting development at the end of the month, still have some major features (devtools, re-eval) I really want to land before returning to contractual work.

Conclusion being basically, currently I do not have the time to look into how webpack dev server works but if anyone wants to look into interop I'll be available to tell them how amok works :smiley:

**Comment by stereokai** [comment]
I'm currently busy over my head... Maybe come the winter, I'll actually have some time to do anything about this...

**Comment by stereokai** [comment]
@caspervonb Needless to say your work is infinitely appreciated.

**Comment by caspervonb** [comment]
Also, #79 might also this redundant, depending on what features you are trying to get out of react-hot-loader.

**Comment by caspervonb** [comment]
> @caspervonb Needless to say your work is infinitely appreciated.
Infinity appreciate that my work is appreciated :grin:

**Comment by stereokai** [comment]
@caspervonb How would #79 make this redundant?

**Comment by caspervonb** [comment]
In short RHL re-evaluates state, but can't deal with closures. Amok deals with closures but can't re-evaluate state. #79 would let you define new properties, re-define primitive values, etc.

**Comment by stereokai** [comment]
For maintaining state of course :)

**Comment by caspervonb** [comment]
Wouldn't amok supersede everything RHL can do at that point tho?

**Comment by stereokai** [comment]
Well, I guess so, if you can define new properties and change values on nested objects - I guess it could definitely serve as a replacement to RHL.

I really want to learn how amok works.

**Comment by caspervonb** [comment]
At the core of amok's hot patching is [v8/liveedit.cc](https://github.com/v8/v8/blob/f0c9cc0bbfd461c7f516799d9a58e9a7395f737e/src/liveedit.cc), it's not *that complex* but it can't be done in user land which is why amok has opted for using the debugger instead of user land `eval`.

> I really want to learn how amok works
Post an(other) issue ;)

**Comment by caspervonb** [comment]
I'll be closing this as currently non actionable. Might be worth visiting after #79 lands to see if there's any value to get from interoperability at that point.

**Comment by caspervonb** [comment]
While I was looking for a more general purpose solution to
the problem, specifically being able to pry open any scope and evaluate within it,
it doesn't seem to be doable at this point in time, at-least not without a lot
of code instrumentation and even then its still kinda 'meh',
and amok isn't the place to be instrumenting code.

So, what is likely to happen, is that when the runtimes have native module we'll be
able to gain better debugger access to module scopes, In node its pretty painless
to monkey patch the runtime to allow this, so while the above is still something
that will eventually go in core but seems non actionable right now.

But, can we do something about bundled modules, right now, today? Possibly.

Lets re-iterate how things work, Amok updates functions in place, without replacing them,
v8's references are kept intact just due to the nature of its being v8 doing
the actual legwork.

The payoff is that we can edit deeply nested functions, like callbacks.
For example, amok's own source code would get pretty great milage out of this,
since it's mostly callback based on http.Server, ws.Server, etc.

Hot module replacements are quite the opposite, as the name implies it replaces
properties on modules, this defeats the whole purpose of amok trying to keep references intact.

However it is not a dead-end, because we can just write a better loader.

**Comment by stereokai** [comment]
I saw that you opened this issue again, and was going to post this:
![image](https://cloud.githubusercontent.com/assets/2685357/9292580/abe3053a-440b-11e5-9a32-222cc640b24b.png)

But you were quicker than me and posted a conclusion as well! Ha.

So what you're basically saying, is, once support for ES6 modules matures across the metaphorical V8 board, there will be debugger APIs to handle them? Maybe I don't understand and native modules doesn't imply ES6 modules. But, otherwise, the question rises regarding userland modules (webpack, browserify, jspm).

> we can just write a better loader

You mean write a loader to "compete" with react-hot-loader, etc?

**Comment by caspervonb** [comment]
> I saw that you opened this issue again, and was going to post this:

What can I say, managing an issue tracker is hard ;)

I closed it as I was realizing the implementation details of such a loader
belongs in its own issue.

> So what you're basically saying, is, once support for ES6 modules matures across
> the metaphorical V8 board, there will be debugger APIs to handle them?
> Maybe I don't understand and native modules doesn't imply ES6 modules.

That's what I am hoping for, don't know anything yet, I try to keep tabs on
v8 commits but haven't really been keeping track lately.

> otherwise, the question rises regarding userland modules (webpack, browserify, jspm).

Thats basically my realization now, that in order to deal with for example
webpack modules, the logical way to do it is via a webpack loader.

> You mean write a loader to "compete" with react-hot-loader, etc?

I said *better*, but I should have said *works better with amok*.

react-hot-loader will work with amok, but they'll be stepping on eachother's toes a little bit.

I'll get into more detail on a seperate issue of the edge cases that needs to
be handled, but basically it will be oriented around adding the things amok
won't touch.

Browserify is still an unsolved issue, but one step at a time.

**Comment by stereokai** [comment]
50% of your comment relates to how difficult it is to keep track of an issue tracker. LOL.

> react-hot-loader will work with amok, but they'll be stepping on eachother's toes a little bit.

Even a little is an issue when the purpose is to maintain a persistent live state between code changes.

> I'll get into more detail on a seperate issue of the edge cases 

That's what I'm talking about! How about having the code that tracks and prepares the new changes in it's own package, and then it could be required by shell webpack- and browserify-plugins?
